Terminal.Gui/Views/ScrollBarView.cs

```diff
@@ -290,14 +199,16 @@ public override bool MouseEvent (MouseEvent mouseEvent)
             return false;
         }
 
-        if (!Host.CanFocus)
+        View host = SuperView is Adornment adornment ? adornment.Parent : SuperView;
```

Comment:
Same comment as above. This is too tightly coupled.

Reply:
That said, this piece of code is central to making all this work. At some point, `ScrollBar` needs to have a protocol with the view whose content is being scrolled.  

What is all the information this protocol needs to convey:

- Focus (both can and has)
- The size of the content.
- The size of the visible window into the content
- ...

What else? Let's get this list flushed out and then design a concise protocol (a set of properties, methods, interfaces, etc...) that cleanly encapsulates them.

Reply:
> Same comment as above. This is too tightly coupled.

This is not a coupling case. It is only to ensure that the Parent is focused if an adornment is clicked if the Parent can be focused.

Reply:
> > Same comment as above. This is too tightly coupled.
> 
> This is not a coupling case. It is only to ensure that the Parent is focused if an adornment is clicked if the Parent can be focused.

I don't see how you can say this is NOT a coupling case. The code above (in `ScrollbarView`) has the word `Adornment` in it. That is coupling. 

I believe it is possible for `ScrollBar` to have NO knowledge of `Adornment`.

Reply:
But the `ScrollBar` need to access the information about the position and size of the parent. Remember the `ScrollBar` now  will be part of the `View` class.
But the fact of the `ScrollBar` can be added on any view sub-class, including adornments, the latest commits broken something.

Reply:
> But the `ScrollBar` need to access the information about the position and size of the parent. Remember the `ScrollBar` now will be part of the `View` class. But the fact of the `ScrollBar` can be added on any view sub-class, including adornments, the latest commits broken something.

Let's be specific about what information the `Parent` needs to provide `ScrollBar`:

1) The position and dimensions of the `ScrollBar`: `ScrollBar.X/Y/Height/Width`.
2) The size of the virtual content: `ScrollBar.Size`
3) The position w/in the virtual content that scrollbar is pointing to: `ScrollBar.Position`
4) The size of the visible area: `Parent.Bounds.Size`.  
5) The offset into the virtual content that the (0,0) of the visible area resides at. `Parent.Bounds.Location`
6) Whether the parent has focus or not. `SuperView.HasFocus`
7) The ColorScheme of the parent. `SuperView.ColorScheme`

I think you're talking about 4 and 5 above because he rest are handled by the fact that `Padding` tracks it's `Parent`, and `ScrollBar` can just reference `SuperView`.

Am I missing anything?

For 4, the size of the visible area, simply define `ScrollBar.VisibleSize` that `Padding` (or `View`) set whenever `Bounds` changes.

For 5, offset into the virtual content that the (0,0) of the visible area resides at, I don't think `ScrollBar` needs this. You currently don't reference `SuperView.Bounds.X/Y` anywhere that I can see.

Reply:
> But the `ScrollBar` need to access the information about the position and size of the parent. Remember the `ScrollBar` now will be part of the `View` class. But the fact of the `ScrollBar` can be added on any view sub-class, including adornments, the latest commits broken something.

IMO, you should not have already deleted the old `ScrollBarView`. Don't try to move all old code that uses it to the new model at once. We can do it over time.

Reply:
Yes I reference `SuperView.Bounds` in various parts of the `ScrollBar`. Look at it better.

Reply:
> IMO, you should not have already deleted the old `ScrollBarView`. Don't try to move all old code that uses it to the new model at once. We can do it over time.

But I already deleted. I only renamed the file and the class name.

Reply:
> > IMO, you should not have already deleted the old `ScrollBarView`. Don't try to move all old code that uses it to the new model at once. We can do it over time.
> 
> But I already deleted. I only renamed the file and the class name.

`git checkout v2_develop ./Terminal.Gui/Views/ScrollBarView.cs ./Terminal.Gui/Views/ScrollView.cs`

Reply:
> Yes I reference `SuperView.Bounds` in various parts of the `ScrollBar`. Look at it better.

I've looked at it closely. A view referencing it's `SuperView.Bounds` is not a problem. A view that knows about `Parent.Bounds` is bad, unless it is a class derived from `Adornment`.

Reply:
> git checkout v2_develop ./Terminal.Gui/Views/ScrollBarView.cs ./Terminal.Gui/Views/ScrollView.cs

But I also renamed `ScrollBarViewTests` to `ScrollBarTests` and I have various changes on both files. If I checkout them from the `v2_develop` branch I'll have a bunch of conflicts, right? I only renamed because you said on a comment.

Reply:
> I've looked at it closely. A view referencing it's `SuperView.Bounds` is not a problem. A view that knows about `Parent.Bounds` is bad, unless it is a class derived from `Adornment`.

Many of the calculations of the `ScrollBar` is based of the size of the `SuperView`. Thus it need to check it to decide it need to scroll or not. With the implementations of negative bounds a ScrollBar thay is on the `ContentArea` will have much more problems to calculate the size because the `SuperView` bounds will have a higher value than the visible `ContentArea` and the `ScrollBar` must only use the visible `ContentArea` size. But because the `ScrollBar` dimensions use `Dim()` or `Dim(1)` it will overflow the visible `ContentArea` size. I hope you can understand this and you may some better suggestion. With `Adornments` this isn't a problem because they aren't use negative bounds.

Reply:
Now I'm thinking that it was the ScrollView that should be renamed ScrollBar and not the ScrollBarView. I will reverse this when this PR is more solid.
See the `ScrollBars` scenari which have `ScrollBar` on all `Adornments` and `ContentArea` with mouse handling. For negative bounds we have to use `Pos/Dim` absolute values to position the `ScrollBar`. I hope you understand why I need to access to the `SuperView`.

Reply:
> Many of the calculations of the ScrollBar is based of the size of the SuperView.

No, this is not strictly true.

The right statement is: 

> Nany of the calculations of the ScrollBar need to use information known by the SuperView when the ScrollBar is used w/in `Padding`." 

In an earlier post I listed exactly what those things (there are 7 of them...unless I'm missing something, which is possible) are. 

1) The position and dimensions of the `ScrollBar`: `ScrollBar.X/Y/Height/Width`.
2) The size of the virtual content: `ScrollBar.Size`
3) The position w/in the virtual content that scrollbar is pointing to: `ScrollBar.Position`
4) The size of the visible area: `Parent.Bounds.Size`.  
5) The offset into the virtual content that the (0,0) of the visible area resides at. `Parent.Bounds.Location`
6) Whether the parent has focus or not. `SuperView.HasFocus`
7) The ColorScheme of the parent. `SuperView.ColorScheme`

These are just data. They can be provided to `ScrollBar` in many different ways. One way is by having `ScrollBar` access `SuperView`, which you keep insisting is the only way.

The way I think will make the new `ScrollBar` simpler and more flexible is to ensure `ScrollBar` supports an API (a set of properties/methods) that let those things be known to `ScrollBar`.  

1, 2, and 3 are already handled this way. 6 and 7 can still work via `SuperView` because that's how most other views work for focus and colorschemes.

For 4: You could add a `public Size VisibleArea` to `ScrollBar`.

For 5: You could add a `public Point ContentOffset` to `ScrollBar`.

Reply:
I already reverted the rename of the `ScrollBarView` and `ScrollBarViewTests`.